src/server.js

```diff
@@ -0,0 +1,245 @@
+import debug from 'debug'
+import WebSocket from 'ws'
+import http from 'http'
+import compose from 'koa-compose'
+import Executor from '@hydre/graphql-batch-executor'
+
+import {
+  on,
+} from 'events'
+
+import {
+  pipeline,
+} from 'stream'
+
+const log = debug('gql-ws').extend('server')
+const DEFAULT_TIMEOUT = 30_000
+
+export default class {
+  #ws_options
+  #schema
+  #rootValue
+  #context
+  #web_server
+
+  #middleware = []
+  #context_is_static
+  #timeout
+
+  /**
+   * Initialize a graphql websocket server with a configuration object
+   * @param {Object} options - server options
+   * @param {Number} [options.timeout=30000] - in millisecond,
+   *  try to deconnect non reachable clients
+   * @param {Object} [options.ws_options] - see https://github.com/websockets/ws/blob/master/doc/ws.md#new-websocketserveroptions-callback
+   * @param {graphql.SchemaDefinitionNode} options.schema - a graphql schema
+   * @param {Object} options.rootValue - a graphql resolver object
+   * @param {(Object|Function)} [options.context={}] - the context object
+   * or as a function/async function to dynamically
+   * build it on every upgrade request
+   */
+  constructor({
+    ws_options = {
+      path             : '/',
+      perMessageDeflate: false,
+      maxPayload       : 500,
+    },
+    schema = (() => {
+      throw new Error('Missing or invalid schema')
+    })(),
+    rootValue = {},
+    context = {},
+    web_server = http.createServer(),
+    timeout = DEFAULT_TIMEOUT,
+  }) {
+    this.#ws_options = ws_options
+    this.#schema = schema
+    this.#rootValue = rootValue
+    this.#web_server = web_server
+    this.#context = context
+    this.#timeout = timeout
+
+    this.#context_is_static = typeof context === 'object'
+  }
+
+  /**
+   * Resolve the context to allow usage of dynamic context building
+   * @param {(Object | Function | AsyncFunction)} context
+   */
+  context(...parameters) {
+    if (this.#context_is_static) return this.#context
+    return this.#context(...parameters)
+  }
+
+  use(middleware) {
+    this.#middleware.push(middleware)
+  }
+
+  listen(...parameters) {
+    const middleware = compose(this.#middleware)
+    const http_server = this.#web_server.listen(...parameters)
+    // we prevent usage of those as it's up to the http server to decide
+    // @see https://github.com/websockets/ws/blob/master/doc/ws.md#new-websocketserveroptions-callback
+    const {
+      // throwing off both vars
+      // eslint-disable-next-line no-unused-vars
+      host,
+      // eslint-disable-next-line no-unused-vars
+      port,
```

Comment:
I think i migth be better to add a `varsIgnorePattern` for the eslint rule like if you variable start with `_`

Reply:
event better i found there is a `ignoreRestSiblings`, added in [`24d92`](https://github.com/usidy/eslint-config-sidy/commit/24d9209ab5f1e14488183089a1595c276142d5f5)